### DISCUSSION TITLE
GetEstimatedChangesCountAsync Server Scope

### DISCUSSION THREAD

**Discussion by dotnetninja** [Mimetis/Dotmim.Sync#615] [Q&A] [answered]
I currently have a webjob that creates a snapshot for a downloadonly scope.  Is it was possible to check to see if any changes exist on the server downloadonly tables before a new downloadonly scope snapshot is created? If so, how is this accomplished? 

I should note this is a multi-scope sync scenario. The snapshot is restricted to download only. The other scope is restricted to true two-way synchronization.

**Comment by Mimetis** [comment, marked as answer]
Hey @dotnetninja 

You can try something like that:

## Get Estimated changes from a client application

If you want to know if any changes, from a client application, (using a `WebClientOrchestrator`) you can do something like that:

``` cs
var webClientOrchestrator = new WebClientOrchestrator(this.ServiceUri);
var agent = new SyncAgent(client.Provider, webClientOrchestrator, options, "scope1");

// Ensure scope is created locally
var clientScope = await agent.LocalOrchestrator.GetClientScopeAsync();

// get changes from server, without any changes sent from client side
var changes = await webClientOrchestrator.GetEstimatedChangesCountAsync(clientScope);

```

## Get Estimated changes from a daemon application running on server side

Well, it's not out of the box, and it could be ... complicated :)

If you want to know it, from a daemon, directly from the server side, it's complicated, because you need the last time your client made a sync, and this information is only available from the client database itself.

We still have this information in the server database, in the history table.
But this information is not guaranteed.
In the history table, all rows correspond to a client (identified by its `sync_scope_id`)
There is no out of the box method (I guess I can make one in the near future ;)

Here is what you can do, from the server side:

``` cs
var serverProvider = new SqlSyncProvider(csServer);

// Remote orchestrator
var remoteOrchestrator = new RemoteOrchestrator(serverProvider, options, setup, "DefaultScope");
// all histories for all client scope
var histories = await remoteOrchestrator.GetServerHistoryScopes();

// get the correct client scope 
var clientScopeId = new Guid("B47C21C2-198D-48CB-9568-7CF1A3333288");

// get the correct history
var history = histories.Where(h => h.Id == clientScopeId).FirstOrDefault();

var clientScope = new ScopeInfo();
clientScope.IsNewScope = false; 
clientScope.Id = clientScopeId;
clientScope.LastServerSyncTimestamp = history.LastSyncTimestamp;
clientScope.Name = "DefaultScope";

var s = await agent.RemoteOrchestrator.GetEstimatedChangesCountAsync(clientScope);
```

Let me know if it's working for you ?

Seb
